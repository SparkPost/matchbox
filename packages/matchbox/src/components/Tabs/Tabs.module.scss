@import '../../styles/config';

.Tabs {
  margin-bottom: spacing();
  padding: 0 rem(18);

  background: color(gray, 10);
  box-shadow: shadow();

  white-space: nowrap;
}

.connectBelow {
  margin-bottom: 0;
  border-top-left-radius: border-radius();
  border-top-right-radius: border-radius();
  border-bottom: 1px solid color(gray, 8);

  & + * {
    border-top-left-radius: 0;
    border-top-right-radius: 0;
  }
}

.fitted {
  display: flex;
  padding: 0;
}

.Tab, a.Tab {
  position: relative;
  display: inline-block;
  padding: rem(15) 0;
  margin-right: rem(21);

  color: color(gray, 4);
  text-decoration: none;

  &:after {
    display: block;
    content: "";
    position: absolute;
    left: 0;
    right: 0;
    bottom: -1px;
    height: rem(4);

    border-radius: 1px;

    opacity: 0;
    transition: 0.15s;
  }

  &:hover, &:active, &:focus {
    color: color(gray, 2);

    &:after {
      background: color(gray, 8);
      opacity: 1;
    }
  }

  &.fittedTab {
    flex: 1 0 0;
    text-align: center;
    margin-right: 0;
  }
}

.selected, a.selected {
  &, &:focus {
    pointer-events: none;

    &:after {
      opacity: 1;
    }

    .color-orange & {
      color: color(orange);
      &:after { background: color(orange); }
    }

    .color-blue & {
      color: color(blue);
      &:after { background: color(blue); }
    }

    .color-navy & {
      color: color(navy);
      &:after { background: color(navy); }
    }

    .color-purple & {
      color: color(purple);
      &:after { background: color(purple); }
    }

    .color-red & {
      color: color(red);
      &:after { background: color(red); }
    }
  }
}
