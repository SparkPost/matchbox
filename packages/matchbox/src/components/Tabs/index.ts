export { default as Tabs } from './Tabs';
export * from './Tabs';
export * from './Tab';
