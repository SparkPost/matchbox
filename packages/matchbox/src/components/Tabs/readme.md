# Tabs

```js
import { Tabs, useTabs } from '@sparkpost/matchbox';

function Example() {
  const {
    getTabsProps, // Provides required tab props, onSelect, selected, and constructed tabs with onClick handlers
    tabIndex, // The selected tab index
    tabs, // The constructed tabs with onClick handlers
    setTabIndex, // Function to programatically set tab index
  } = useTabs({
    tabs, // Your tabs array
    initialSelected, // Sets initial selected tab
  });

  return (
    <>
      <Tabs
        {...getTabProps()}
        // When set to 'auto', automatically activates tabs when using focus keyboard navigation
        // When set to 'manual', tab activation requires using space or enter keys after focusing
        keyboardActivation="manual"
        // Stretches the tabs to span the full width
        fitted
        // System prop to manually overwrite border
        borderBottom="none"
        // System prop to manually set margins
        m="500"
      />
    </>
  );
}
```
