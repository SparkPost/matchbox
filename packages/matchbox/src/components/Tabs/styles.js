import { buttonReset } from '../../styles/helpers';
import { tokens } from '@sparkpost/design-tokens';

export const tabStyles = ({ selected, fitted }) => `
  ${buttonReset}
  position: relative;
  flex: ${fitted ? '1' : '0'};
  
  padding: 0 ${tokens.spacing_200};
  ${'' /* Not a token, equivalent to 20px to enqure 10rem of spacing between text */}
  margin: 0 ${!fitted ? '1.25rem' : '0'};

  font-size: ${tokens.fontSize_200};
  font-weight: ${tokens.fontWeight_medium};
  line-height: 3.75rem; ${'' /* Equivalent to 60px */}
  color: ${selected ? tokens.color_blue_700 : tokens.color_gray_700};

  &:after {
    display: block;
    position: absolute;
    content: '';
    bottom: -1px;
    left: 0;
    right: 0;
    height: ${tokens.spacing_100};
    background: ${selected ? tokens.color_blue_700 : 'transparent'};
    transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
  }

  &:hover {
    ${!selected ? `color: ${tokens.color_gray_900};` : ''};
    &:after {
      ${!selected ? `background: ${tokens.color_gray_400};` : ''}
    }
  }
`;

export const wrapperStyles = () => `
  display: flex;
  border-bottom: ${tokens.borderWidth_100} solid ${tokens.color_gray_400};
`;
