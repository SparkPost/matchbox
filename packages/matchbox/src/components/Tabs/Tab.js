import React from 'react';
import styled from 'styled-components';
import { UnstyledLink } from '../UnstyledLink';
import { tabStyles } from './styles';

// TODO Replace this when styled-components supports shouldForwardProps
// See: https://github.com/styled-components/styled-components/commit/e02109e626ed117b76f220d0b9b926129655262d
// Or when UnstyledLink is updated to use system props
const LinkWrapper = React.forwardRef(function LinkWrapper(props, ref) {
  const { selected, fitted, ...rest } = props;
  return <UnstyledLink ref={ref} {...rest} />;
});

const StyledTab = styled(LinkWrapper)`
  ${tabStyles}
`;

const Tab = React.forwardRef(function Tab(props, ref) {
  const { index, content, selected, fitted, component, Component, tabIndex, type, ...rest } = props;

  function handleClick(event) {
    const { index, onClick } = props;
    onClick(event, index);
  }

  // Buttons ensure focusability
  // Links will be focusable with an href
  // TODO deprecate `Component`
  const wrapper = component || Component || 'button';

  return (
    <StyledTab
      aria-selected={selected === index}
      as={wrapper}
      selected={selected === index}
      fitted={fitted}
      ref={ref}
      {...rest}
      onClick={handleClick}
      role="tab"
      tabIndex={tabIndex}
      // Safari overwrites the button reset styles
      // if the element is an 'a' with this attribute
      type={wrapper === 'button' ? 'button' : type}
    >
      {content}
    </StyledTab>
  );
});

Tab.displayName = 'Tab';

export default Tab;
