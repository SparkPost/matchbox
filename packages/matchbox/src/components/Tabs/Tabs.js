import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { UnstyledLink } from '../UnstyledLink';

import styles from './Tabs.module.scss';

class Tab extends Component {
  static displayName = 'Tab';

  handleClick = (event) => {
    const { index, onClick } = this.props;
    onClick(event, index);
  }

  render() {
    const { index, content, selected, fittedTab, ...rest } = this.props;
    const classes = classnames(
      styles.Tab,
      selected === index && styles.selected,
      fittedTab && styles.fittedTab
    );

    return (
      <UnstyledLink className={classes} {...rest} onClick={this.handleClick}>
        {content}
      </UnstyledLink>
    );
  }
}

class Tabs extends Component {
  static displayName = 'Tabs';

  static propTypes = {
    /**
     * Tab Content
     * Actions that build the tabs. Most button and unstyled link props will work in here.
     * e.g. { content: 'Label', onClick: callback() }
     */
    tabs: PropTypes.arrayOf(PropTypes.shape({
      content: PropTypes.node.isRequired
    })),

    /**
     * Tab Color
     */
    color: PropTypes.oneOf(['orange', 'blue', 'navy', 'purple', 'red']),
    /**
      * Index of selected tab
      */
    selected: PropTypes.number.isRequired,
    /**
      * Connects this component with component underneath it. Works well with Panels.
      */
    connectBelow: PropTypes.bool,

    onSelect: PropTypes.func
  };

  static defaultProps = {
    connectBelow: true,
    color: 'orange'
  };

  handleClick = (event, index) => {
    const { onSelect, selected, tabs } = this.props;
    const { onClick } = tabs[index];

    if (onClick) {
      onClick(event, index);
    }

    if (onSelect && selected !== index) {
      onSelect(index, selected);
    }
  }

  render() {
    const { tabs, selected, connectBelow, color, fitted } = this.props;

    const tabMarkup = tabs.map((tab, i) => (
      <Tab key={i} index={i} fittedTab={fitted} selected={selected} {...tab} onClick={this.handleClick} />
    ));

    const tabsClasses = classnames(
      styles.Tabs,
      styles[`color-${color}`],
      connectBelow && styles.connectBelow,
      fitted && styles.fitted
    );

    return (
      <div className={tabsClasses}>
        {tabMarkup}
      </div>
    );
  }
}

export default Tabs;
