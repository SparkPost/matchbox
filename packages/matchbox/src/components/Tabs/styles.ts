import { buttonReset, focusOutline } from '../../styles/helpers';
import { tokens } from '@sparkpost/design-tokens';

export const tabStyles = ({
  $selected,
  $fitted,
}: {
  $selected?: boolean;
  $fitted?: boolean;
}): string => `
  ${buttonReset}
  
  ${focusOutline({ offset: '0px', modifier: '&' })}
  &:focus:not(:focus-visible):after {
    box-shadow: none;
  }
  ${focusOutline({ offset: '0px', modifier: '&:focus-visible' })}

  display: inline-block;
  cursor: pointer;
  position: relative;
  flex: ${$fitted ? '1' : '0'};
  text-decoration: none;
  outline: none;
  text-align: center;

  padding: 0 ${tokens.spacing_400};
  margin-left: ${$fitted ? '0' : tokens.sizing_200};

  &:first-child {
    margin-left: 0;
  }

  font-size: ${tokens.fontSize_300};
  font-weight: ${tokens.fontWeight_medium};
  line-height: ${tokens.sizing_650};
  white-space: nowrap;

  background: ${$selected ? tokens.color_blue_200 : 'transparent'};
  transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};

  &, &:visited {
    color: ${$selected ? tokens.color_blue_700 : tokens.color_gray_700};
  }

  &:before {
    display: block;
    position: absolute;
    content: '';
    bottom: 0px;
    left: 0;
    right: 0;
  }

  &:before {
    height: 1px;
    background: ${$selected ? tokens.color_blue_700 : 'transparent'};
    transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
  }
  
  &:hover {
    ${
      !$selected
        ? `
      color: ${tokens.color_gray_900};
      background: ${tokens.color_gray_200};
    `
        : ''
    };
    &:before {
      ${!$selected ? `background: ${tokens.color_gray_700};` : ''}
    }
  }
  
  &:after {
    top: 2px;
  }
`;

export const overflowTabs = ({ isOverflowing }: { isOverflowing?: boolean }): string => `
  position: absolute;
  left: 0;
  top: 0;
  right:0;
  display: flex;
  visibility: ${isOverflowing ? 'hidden' : 'visible'};
  pointer-events: ${isOverflowing ? 'none' : 'auto'};
  overflow: ${isOverflowing ? 'hidden' : 'unset'};
`;

export const containerStyles = (): string => `
  position: relative;
  height: ${tokens.sizing_650};
`;
