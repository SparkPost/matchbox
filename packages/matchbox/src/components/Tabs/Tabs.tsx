import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { MoreHoriz } from '@sparkpost/matchbox-icons';
import { margin, borderBottom } from 'styled-system';
import { createPropTypes } from '@styled-system/prop-types';
import { pick } from '@styled-system/props';
import { ActionList } from '../ActionList';
import { Box } from '../Box';
import { Button } from '../Button';
import { Popover } from '../Popover';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { deprecate } from '../../helpers/propTypes';
import useTabConstructor from './useTabConstructor';
import Tab, { TabProps } from './Tab';
import { containerStyles, overflowTabs } from './styles';
import { useResizeObserver } from '../../hooks';

import { getPositionFor } from '../../helpers/geometry';

const OverflowTabContainer = styled('div')`
  ${overflowTabs}
`;

const Container = styled('div')`
  ${margin}
  ${borderBottom}
  ${containerStyles}
`;

export type TabsProps = {
  disableResponsiveBehavior?: boolean;
  keyboardActivation?: 'auto' | 'manual';
  tabs?: TabProps[];
  selected: number;
  onSelect?: (index: number) => void;
  fitted?: boolean;
};

const Tabs = React.forwardRef<HTMLDivElement, TabsProps>(function Tabs(props, userRef) {
  const {
    disableResponsiveBehavior,
    keyboardActivation = 'auto',
    tabs,
    selected,
    onSelect,
    fitted,
    ...rest
  } = props;
  const [isOverflowing, setIsOverflowing] = React.useState(false);
  const [popoverOpen, setPopoverOpen] = React.useState(false);

  const wrapperRef = React.useRef();
  const overflowRef = React.useRef();

  const [resizeRef, { contentRect }] = useResizeObserver();

  function handleClick(event: React.MouseEvent<HTMLAnchorElement>, index: number) {
    const { onClick } = tabs[index];

    if (onClick) {
      onClick(event, index);
    }

    if (onSelect && selected !== index) {
      onSelect(index);
    }

    if (isOverflowing) {
      setPopoverOpen(false);
    }
  }

  React.useLayoutEffect(() => {
    if (!disableResponsiveBehavior) {
      const wrapperPosition = getPositionFor(wrapperRef.current);
      const overflowPosition = getPositionFor(overflowRef.current);

      // Compares tab container width with its position relative to a measurement pixel
      if (wrapperPosition.width < overflowPosition.left - wrapperPosition.left) {
        setIsOverflowing(true);
      } else {
        setIsOverflowing(false);
      }
    }
  }, [wrapperRef, overflowRef, contentRect, tabs]);

  // Constructs the tabs, their props and handles tab keyboard navigation
  const { tabMarkup, tabActions, focusContainerProps } = useTabConstructor({
    tabs,
    fitted,
    handleClick,
    selected,
    onSelect,
    keyboardActivation,
    disableResponsiveBehavior,
  });

  const assignWrapperRefs = (node) => {
    if (wrapperRef) {
      wrapperRef.current = node;
    }
    if (userRef) {
      (userRef as React.MutableRefObject<HTMLDivElement>).current = node;
    }
    if (resizeRef) {
      resizeRef(node);
    }
  };

  return (
    <Container borderBottom="400" {...pick(rest)} ref={assignWrapperRefs} tabIndex="-1">
      <Box aria-hidden={isOverflowing}>
        <OverflowTabContainer
          aria-orientation="horizontal"
          isOverflowing={isOverflowing}
          role="tablist"
        >
          <Box
            position="absolute"
            left="0"
            top="0"
            right="0"
            display="flex"
            {...focusContainerProps}
          >
            {tabMarkup}
            {/* Measurement pixel used to detect content overflow */}
            <Box display="inline-block" width="1px" height="1px" ref={overflowRef} />
          </Box>
        </OverflowTabContainer>
      </Box>
      {isOverflowing && !disableResponsiveBehavior && (
        <Box display="flex" alignItems="center" justifyContent="space-between">
          <Box flex="0">
            <Tab index={selected} selected={selected} {...tabs[selected]} />
          </Box>

          <Box flex="0">
            <Popover
              id="tab-options"
              left
              bottom
              open={popoverOpen}
              onClose={() => setPopoverOpen(false)}
              trigger={
                <Button
                  aria-controls="tab-options"
                  aria-expanded={popoverOpen}
                  data-id="tab-options-button"
                  variant="text"
                  onClick={() => setPopoverOpen(!popoverOpen)}
                >
                  <MoreHoriz size={20} />
                  <ScreenReaderOnly>More Options</ScreenReaderOnly>
                </Button>
              }
            >
              <ActionList>
                {tabActions.map((action, i) => (
                  <ActionList.Action key={i} {...action} />
                ))}
              </ActionList>
            </Popover>
          </Box>
        </Box>
      )}
    </Container>
  );
});

Tabs.displayName = 'Tabs';
Tabs.propTypes = {
  disableResponsiveBehavior: PropTypes.bool,
  /**
   * Tab Content
   * Actions that build the tabs. Most button and unstyled link props will work in here.
   * e.g. { content: 'Label', onClick: callback() }
   */
  tabs: PropTypes.arrayOf(
    PropTypes.shape({
      content: PropTypes.node.isRequired,
    }),
  ),

  /**
   * Tab Color
   */
  color: deprecate(
    PropTypes.oneOf(['orange', 'blue', 'navy', 'purple', 'red']),
    'Tab color is no longer configurable',
  ),
  keyboardActivation: PropTypes.oneOf(['auto', 'manual']),
  /**
   * Index of selected tab
   */
  selected: PropTypes.number.isRequired,
  onSelect: PropTypes.func,
  ...createPropTypes(margin.propNames),
  ...createPropTypes(borderBottom.propNames),
};

export default Tabs;
