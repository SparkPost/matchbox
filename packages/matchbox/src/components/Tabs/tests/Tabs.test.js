import React from 'react';
import { shallow } from 'enzyme';

import Tabs from '../Tabs';

describe('Tabs', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      tabs: [
        { content: 'Tab 1', onClick: jest.fn() },
        { content: 'Tab 2', onClick: jest.fn() }
      ],
      selected: 0,
      connectBelow: false,
      onSelect: jest.fn()
    };

    wrapper = shallow(<Tabs {...props}/>);
  });

  it('renders with first tab selected', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Tab').at(0).dive()).toMatchSnapshot();
    expect(wrapper.find('Tab').at(1).dive()).toMatchSnapshot();
  });

  it('renders with fitted tabs', () => {
    wrapper.setProps({ fitted: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders when connected with below components', () => {
    wrapper.setProps({ connectBelow: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('changes selected tab on click', () => {
    const event = {};
    wrapper.find('Tab').at(1).simulate('click', event, 1);
    expect(props.tabs[1].onClick).toHaveBeenCalledTimes(1);
    expect(props.onSelect).toHaveBeenCalledWith(1, 0);
  });
});
