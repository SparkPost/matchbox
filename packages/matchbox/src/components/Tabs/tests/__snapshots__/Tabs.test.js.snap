// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders when connected with below components 1`] = `
<div
  className="Tabs color-orange connectBelow"
>
  <Tab
    content="Tab 1"
    index={0}
    key="0"
    onClick={[Function]}
    selected={0}
  />
  <Tab
    content="Tab 2"
    index={1}
    key="1"
    onClick={[Function]}
    selected={0}
  />
</div>
`;

exports[`Tabs renders with first tab selected 1`] = `
<div
  className="Tabs color-orange"
>
  <Tab
    content="Tab 1"
    index={0}
    key="0"
    onClick={[Function]}
    selected={0}
  />
  <Tab
    content="Tab 2"
    index={1}
    key="1"
    onClick={[Function]}
    selected={0}
  />
</div>
`;

exports[`Tabs renders with first tab selected 2`] = `
<UnstyledLink
  className="Tab selected"
  onClick={[Function]}
>
  Tab 1
</UnstyledLink>
`;

exports[`Tabs renders with first tab selected 3`] = `
<UnstyledLink
  className="Tab"
  onClick={[Function]}
>
  Tab 2
</UnstyledLink>
`;

exports[`Tabs renders with fitted tabs 1`] = `
<div
  className="Tabs color-orange fitted"
>
  <Tab
    content="Tab 1"
    fittedTab={true}
    index={0}
    key="0"
    onClick={[Function]}
    selected={0}
  />
  <Tab
    content="Tab 2"
    fittedTab={true}
    index={1}
    key="1"
    onClick={[Function]}
    selected={0}
  />
</div>
`;
